pyroscope/symb/table/table.go

```diff
@@ -0,0 +1,216 @@
+
+	log "github.com/sirupsen/logrus"
+	"go.opentelemetry.io/ebpf-profiler/pyroscope/symb/ffi"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	magic   uint32 = 0x6c627467 // "gtbl"
+	version uint32 = 1
+)
+
+var (
+	versionName = fmt.Sprintf("table-%d", version)
+)
+
+func VersionName() string {
+	return versionName
+}
+
+type header struct {
+	magic         uint32
+	version       uint32
+	rangesOffset  uint64
+	stringsOffset uint64
+	padding       uint64
+}
+
+type entry struct {
+	va     uint32
+	length uint32
+	depth  uint32
+	fun    uint32
+}
+
+func (e entry) String() string {
+	return fmt.Sprintf("va: %x, length: %d, depth: %d, fun: %d", e.va, e.length, e.depth, e.fun)
+}
+
+// todo add 64bit support
+// todo change the layout of the file, separate tables for addresses and the rest of entry fields ?
+type Table struct {
+	file         *os.File
+	rangesOffset uint64
+	rangesCount  uint64
+	strOff       uint64
+}
+
+func OpenPath(path string) (*Table, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	return OpenFile(f)
+}
+
+func OpenFile(f *os.File) (*Table, error) {
+	var err error
+	res := new(Table)
+
+	res.file = f
+
+	headerBuf := make([]byte, unsafe.Sizeof(header{}))
+	if _, readErr := res.file.Read(headerBuf); readErr != nil {
+		res.Close()
+		return nil, readErr
+	}
+
+	hdr := (*header)(unsafe.Pointer(&headerBuf[0]))
+
+	if hdr.magic != magic {
+		res.Close()
+		return nil, errors.New("invalid magic number")
+	}
+	if hdr.version != version {
+		res.Close()
+		return nil, errors.New("unsupported version")
+	}
+
+	res.rangesOffset = hdr.rangesOffset
+	res.rangesCount = (hdr.stringsOffset - hdr.rangesOffset) / 16 // each entry is 16 bytes
+	res.strOff = hdr.stringsOffset
+
+	err = unix.Fadvise(int(res.file.Fd()), 0, 0, unix.FADV_RANDOM)
+	if err != nil {
+		fmt.Printf("failed to Fadvise: %s\n", err)
+	}
+
+	runtime.SetFinalizer(res, func(r *Table) {
+		if r.file != nil {
+			fmt.Printf("WARNING: unclosed file %s\n", res.file.Name())
+			r.file.Close()
+		}
+	})
+	return res, nil
+}
+
+func (st *Table) getEntry(i int) (entry, error) {
+	if i < 0 || uint64(i) >= st.rangesCount {
+		return entry{}, errors.New("index out of bounds")
+	}
+
+	entrySize := int64(unsafe.Sizeof(entry{}))
+	offset := int64(st.rangesOffset) + int64(i)*entrySize
+
+	buf := make([]byte, entrySize)
+	_, err := st.file.ReadAt(buf, offset)
+	if err != nil {
+		return entry{}, err
+	}
+
+	return *(*entry)(unsafe.Pointer(&buf[0])), nil
+}
+
+func (st *Table) Close() {
+	if st.file != nil {
+		_ = st.file.Close()
+	}
+}
+
+func (st *Table) function(offset uint32) string {
+	var strLen uint32
+	buf := make([]byte, 4)
+	if _, err := st.file.ReadAt(buf, int64(st.strOff+uint64(offset))); err != nil {
+		return ""
+	}
+	strLen = binary.LittleEndian.Uint32(buf)
+
+	strData := make([]byte, strLen)
+	if _, err := st.file.ReadAt(strData, int64(st.strOff+uint64(offset)+4)); err != nil {
+		if err != io.EOF {
+			return ""
+		}
+	}
+	return string(strData)
+}
```

Comment:
Here and elsewhere else: we do not verify data validity. A bit flip will cause major issues: 

```go
	if _, err := st.file.ReadAt(buf, int64(st.strOff+uint64(offset))); err != nil {
		return ""
	}
	strLen = binary.LittleEndian.Uint32(buf)

	strData := make([]byte, strLen)
```

`strLen` may be an arbitrary value.

Use of CRC is particularly difficult because we do not have "frames", "blocks", or "pages". I don't know how we could circumvent this.

I would probably split ranges into blocks/pages – that would:
 1. Greatly improve IO (though, this might be implemented via mmap + huge pages).
 2. Allow for check sums.

Same can be achieved by loading the whole `ranges` array into memory (as you only need to calculate the checksum once)